tests/cases/conformance/types/members/indexSignatures1.ts

```diff
@@ -0,0 +1,210 @@
+// @strict: true
+// @declaration: true
+// @target: esnext
+
+// Symbol index signature checking
+
+const sym = Symbol();
+
+function gg3(x: { [key: string]: string }, y: { [key: symbol]: string }, z: { [sym]: number }) {
+    x = z;
+    y = z;  // Error
+}
+
+// Overlapping index signatures
+
+function gg1(x: { [key: `a${string}`]: string, [key: `${string}a`]: string }, y: { [key: `a${string}a`]: string }) {
+    x = y;
+    y = x;
+}
+
+interface IX { [key: `a${string}`]: string, [key: `${string}a`]: string }
+interface IY { [key: `a${string}a`]: string }
+
+function gg2(x: IX, y: IY) {
+    x = y;  // Error
+    y = x;
+}
+
+// Intersection of multiple applicable index signatures
+
+declare let combo: { [x: `foo-${string}`]: 'a' | 'b' } & { [x: `${string}-bar`]: 'b' | 'c' };
+const x1 = combo['foo-test'];  // 'a' | 'b'
+const x2 = combo['test-bar'];  // 'b' | 'c'
+const x3 = combo['foo-test-bar'];  // 'b' (('a' | 'b') & ('b' | 'c'))
+
+declare var str: string;
+
+const x4 = combo[`foo-${str}`];
+const x5 = combo[`${str}-bar`];
+const x6 = combo[`foo-${str}-bar`];
+
+declare let combo2: { [x: `${string}xxx${string}` & `${string}yyy${string}`]: string };
+
+const x7 = combo2['axxxbyyyc'];
+const x8 = combo2['ayyyxxxbc'];
+const x9 = combo2['axxxbbbyc'];  // Error
+
+// Property access on template pattern index signature
+
+declare let dom: { [x: `data${string}`]: string };
+const y1 = dom['data123'];
+const y2 = dom.data123;
+
+// Excess property checking for template pattern index signature
+
+dom = { data123: 'hello' };
+dom = { date123: 'hello' };  // Error
+
+// Contextual typing by index signature with template literal pattern
+
+type Funcs = {
+    [key: `s${string}`]: (x: string) => void,
+    [key: `n${string}`]: (x: number) => void,
+}
+
+const funcs: Funcs = {
+    sfoo: x => x.length,  // x: string
+    nfoo: x => x * 2,     // n: number
+}
+
+// Duplicate index signature checking
+
+type Duplicates = {
+    [key: string | number]: any;  // Error
+    [key: number | symbol]: any;  // Error
+    [key: symbol | `foo${string}`]: any;  // Error
+    [key: `foo${string}`]: any;  // Error
+}
+
+// Conflicting index signature checking
+
+type Conflicting = {
+    [key: `a${string}`]: 'a';
+    [key: `${string}a`]: 'b';
+    [key: `a${string}a`]: 'c';  // Error
+}
+
+// Invalid index signatures
+
+type Invalid<T extends string> = {
+    [key: 'a' | 'b' | 'c']: string;  // Error
+    [key: T | number]: string;  // Error
+    [key: Error]: string;  // Error
+    [key: T & string]: string;  // Error
+}
+
+// Intersections in index signatures
+
+type Tag1 = { __tag1__: void };
+type Tag2 = { __tag2__: void };
+
+type TaggedString1 = string & Tag1;
+type TaggedString2 = string & Tag2;
+
+declare let obj1: { [key: TaggedString1]: string };
+declare let obj2: { [key: TaggedString2]: string };
+declare let obj3: { [key: TaggedString1 | TaggedString2]: string };
+declare let obj4: { [key: TaggedString1 & TaggedString2]: string };

```

Comment:
Can we add some lines to this test where we relate these objects to one another (rather than just accessing them) - index signature relating is a special codepath, so it's worth checking that, eg, union and intersection index signatures still have the correct relationships with both one another and with the `number` and `string` index signatures (for unions/intersections of tagged `number`s and `string`s).